doc/user-guide.md

```diff
@@ -1 +1,380 @@
-# TODO
+# User Guide for `git-meta`
+
+The `git-meta` plugin facilitates implementing a *mono-repo* in terms of a Git
+*meta-repo* and its associated *sub-repos*.  It is a command-line extension of
+Git that allows normal Git operations to be performed on a meta-repo as if it
+were a single mono-repo rather than a collection of (potentially closed)
+submodules.  See the [architecture](./architecture.md) document for more
+detail and definition of these terms.
+
+In this document we first describe the underlying mission and philosophy of
+git-meta.  Then, we explore the types of functionality it offers.  Next, we
+describe the client-side model that git-meta uses to map a logical mono-repo
+onto a meta-repo with its sub-repos.  Finally, we provide a mini-tutorial with
+basic usage scenarios.
+
+## Mission
+
+We intend for git-meta to be _sugar_ that makes it *easy*, but not *possible*
+to use a mono-repo: all operations necessary for mono-repo workflows can be
+performed (perhaps inconveniently) with "plain" Git.  Some related
+requirements:
+
+1. git-meta does not need new meta information unknown to Git to describe the
+   state of a repository.
+1. A repository that is in a valid state according to Git is understandable to
+   git-meta.
+1. A repository that is in a valid state according to git-meta is in a valid
+   state according to Git.
+1. Operations performed by git-meta can be rewritten as git operations.
+
+## Functionality
+
+The functionality of git-meta can be decomposed into four categories:
+
+1. _new behavior_ -- required for implementation of the mono-repo
+1. _submodule boundary erasure_ -- to eliminate the boundaries between
+   submodules and provide for seamless interaction with the (checked out
+   portions of a) mono-repo tree
+1. _better submodule interactions_ -- where existing Git commands, while
+   technically correct,  are counter-intuitive when used with submodules in the
+   context of a mono-repo
+1. _workarounds_ -- for Git commands that are broken or scale poorly with large
+   numbers of submodules
+
+We provide more and examples of these categories below.  The long-term plan is
+to upstream (into Git and libgit2) changes to make the latter two categories
+unnecessary.
+
+### New Behavior
+
+Interacting with a mono-repo (as discussed in the
+[architecture](./architecture.md) document) imposes two client-side
+requirements:
+
+1. When pushing changes, synthetic-meta-refs must be pushed into each affected
+   submodule before pushing the connecting meta-repo commit.
+2. Shas must be fetched on-demand for open submodules during many operations,
+   such as when HEAD is changed in the meta repo.
+
+For the first requirement, we provide `git meta push`:
+
+```bash
+$ git meta push origin master:master
```

Comment:
In the Quick Start, we didn't put a dest ref since it matched the src. We should probably be consistent with that